File: src/main/java/ru/sbt/ss/quicksort/SequentialQuickSort.java
package ru.sbt.ss.quicksort;

import java.util.Random;

public class SequentialQuickSort {

    public static void quickSort(int[] arr, int begin, int end) {
        if (end - begin <= 1) {
            return;
        }
        int m = partition(arr, begin, end);
        quickSort(arr, begin, m);
        quickSort(arr, m, end);
    }

    static int partition(int[] arr, int begin, int end) {
        Random random = new Random();
        int pivotInd = random.nextInt(begin, end);
        int pivot = arr[pivotInd];
        var p1 = begin;
        var p2 = end - 1;
        while (p1 <= p2) {
            while (arr[p1] < pivot) p1++;
            while (arr[p2] > pivot) p2--;
            if (p1 <= p2) swap(arr, p1++, p2--);
        }

        return p1;
    }
    private static void swap(int[] arr, int i, int end) {
        int swapTemp = arr[i];
        arr[i] = arr[end];
        arr[end] = swapTemp;
    }
}

File: src/main/kotlin/ru/sbt/ss/quicksort/sort/SequentialQuickSort.kt
package ru.sbt.ss.quicksort.sort

import ru.sbt.ss.quicksort.array.swap
import kotlin.random.Random

fun IntArray.sequentialQuickSort() {
    sequentialQuickSort(0, size)
}

internal fun IntArray.sequentialQuickSort(l: Int, r: Int) {
    if (r - l <= 1) {
        return
    }
    val m = makePartition(l, r)
    sequentialQuickSort(l, m)
    sequentialQuickSort(m, r)
}

internal fun IntArray.makePartition(l: Int, r: Int): Int {
    val xIndex = Random.nextInt(l, r)
    val x = this[xIndex]
    var p1 = l
    var p2 = r - 1
    while (p1 <= p2) {
        while (this[p1] < x) p1++
        while (this[p2] > x) p2--
        if (p1 <= p2) swap(p1++, p2--)
    }
    return p1
}